parachain/pallets/outbound-queue/src/lib.rs

```diff
@@ -272,6 +279,28 @@ pub mod pallet {
 			let enqueued_message: EnqueuedMessage =
 				EnqueuedMessage::decode(&mut message).map_err(|_| ProcessMessageError::Corrupt)?;
 
+			let halt_check_required = match enqueued_message.command {
+				Command::Upgrade { .. } => false,
+				Command::SetOperatingMode { .. } => false,
+				_ => true,
+			};
+			if halt_check_required {
+				Self::ensure_not_halted().map_err(|_| ProcessMessageError::Yield)?;
+			}
+
+			// Yield low priority message if there was pending high priority message
+			let high_priority = enqueued_message.origin == T::OwnParaId::get();
```

Comment:
The problem here is that the message needs to be decoded to determine the origin. This is too expensive.

See my other latest comment about the origin passed to ProcessMessage::process_message. That's the one you should check.